src/core/mac/mac_frame.cpp

```diff
@@ -769,104 +706,125 @@ void Frame::SetPayloadLength(uint16_t aLength)
 
 uint8_t Frame::SkipSecurityHeaderIndex(void) const
 {
-    uint8_t  index = SkipAddrFieldIndex();
-    uint16_t fcf;
+    uint8_t index = SkipAddrFieldIndex();
 
     VerifyOrExit(index != kInvalidIndex, OT_NOOP);
-    fcf = GetFrameControlField();
 
-    // Security Control + Frame Counter + Key Identifier
-    if ((fcf & kFcfSecurityEnabled) != 0)
+    if ((GetFrameControlField() & kFcfSecurityEnabled) != 0)
     {
+        uint8_t securityControl;
+
         VerifyOrExit(index < GetPsduLength(), index = kInvalidIndex);
-        uint8_t securityControl = *(GetPsdu() + index);
+        securityControl = *(GetPsdu() + index);
 
-        index += kSecurityControlSize + kFrameCounterSize;
+        index += CalculateSecurityHeaderSize(securityControl);
+    }
 
-        switch (securityControl & kKeyIdModeMask)
-        {
-        case kKeyIdMode0:
-            index += kKeySourceSizeMode0;
-            break;
+exit:
+    return index;
+}
 
-        case kKeyIdMode1:
-            index += kKeySourceSizeMode1 + kKeyIndexSize;
-            break;
+uint8_t Frame::CalculateSecurityHeaderSize(uint8_t aSecurityControl)
+{
+    uint8_t size = 0;
 
-        case kKeyIdMode2:
-            index += kKeySourceSizeMode2 + kKeyIndexSize;
-            break;
+    if (aSecurityControl & kSecLevelMask)
+    {
+        size += kSecurityControlSize + kFrameCounterSize;
+    }
 
-        case kKeyIdMode3:
-            index += kKeySourceSizeMode3 + kKeyIndexSize;
-            break;
-        }
+    switch (aSecurityControl & kKeyIdModeMask)
+    {
+    case kKeyIdMode0:
+        size += kKeySourceSizeMode0;
+        break;
+
+    case kKeyIdMode1:
+        size += kKeySourceSizeMode1 + kKeyIndexSize;
+        break;
+
+    case kKeyIdMode2:
+        size += kKeySourceSizeMode2 + kKeyIndexSize;
+        break;
+
+    case kKeyIdMode3:
+        size += kKeySourceSizeMode3 + kKeyIndexSize;
+        break;
     }
 
+    return size;
+}
+
+uint8_t Frame::SkipAddrFieldIndex(void) const
+{
+    uint8_t index;
+
+    VerifyOrExit(kFcfSize + kDsnSize + GetFcsSize() <= GetPsduLength(), index = kInvalidIndex);
+
+    index = CalculateAddrFieldSize(GetFrameControlField());
 exit:
     return index;
 }
 
-uint8_t Frame::SkipAddrFieldIndex(void) const
+uint8_t Frame::CalculateAddrFieldSize(uint16_t aFcf)
 {
-    uint8_t  index = kFcfSize + kDsnSize; // Frame Control field and Sequence Number field
-    uint16_t fcf;
+    uint8_t size = kFcfSize + kDsnSize;
 
-    VerifyOrExit((index + GetFcsSize()) <= GetPsduLength(), index = kInvalidIndex);
-    fcf = GetFrameControlField();
+    // This static method calculates the size (number of bytes) of
+    // Address header filed for a given Frame Control `aFcf` value.
+    // The size includes the Frame Control and Sequence Number fields
+    // along with Destination and Source PAN ID and Short/Extended
```

Comment:
What about only excluding the `Frame Control` and `Sequence Number` from this size? 
* First I think in this way its behavior fits its name more. 
* Secondly this function currently has the same behavior with `SkipAddrFieldIndex`, the only difference is one takes a parameter `aFcf`. And I think this is different from the case, for example, `IsVersion2015()` has two versions, where `IsVersion2015(void)` is public and will be used externally. Both `SkipAddrFieldIndex` and `CalculateAddrFieldSize` are only used internally. So I think we can make them for different behaviors.